### PR TITLE
bpo-30152: Reduce the number of imports for argparse.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue30152

### CODE REVIEW

File: Lib/collections/__init__.py [hunk 1]
```diff
@@ -551,7 +550,8 @@ def most_common(self, n=None):
         # Emulate Bag.sortedByCount from Smalltalk
         if n is None:
             return sorted(self.items(), key=_itemgetter(1), reverse=True)
-        return _heapq.nlargest(n, self.items(), key=_itemgetter(1))
+        import heapq
+        return heapq.nlargest(n, self.items(), key=_itemgetter(1))
```
Comment (rhettinger) [inline Lib/collections/__init__.py]: I don't think this change is reasonable.  You're micro-optimizing the expense of writing normal maintainable code.

**Comment by nedbat** [comment]
People read the stdlib expecting to find best practices.  As the discussion on bpo makes clear, these changes are controversial.  Could we at least have some comments explaining the unusual lines of code (imports inside functions).

**Comment by wm75** [comment]
Regarding the import of copy, I'm not quite sure why it is used at all in the AppendAction/AppendConstAction. Couldn't it be replaced with a simple slice copy of the lists?

**Comment by serhiy-storchaka** [comment]
The default value can be not a list. It can be a deque or a custom sequence with special append method (see an example in https://bugs.python.org/issue16399).

**Comment by wm75** [comment]
I see, thanks. What may be possible though is to make the import of copy even more conditional by first trying a slice copy (which I guess would succeed in > 90% of cases), and resort to copy.copy only if that raises. Maybe needing copy.copy only in *very* exotic cases makes the unconventional inside-a-function import a bit more justifiable?

**Comment by serhiy-storchaka** [comment]
Good idea. Thanks @wm75.

File: Lib/argparse.py [hunk 2]
```diff
@@ -137,10 +134,16 @@ def _get_args(self):
...
+def _copy_items(items):
+    if items is None:
+        return []
+    # The copy module is used only in the 'append' and 'append_const'
+    # actions, and it is needed only when the default value isn't a list.
+    # Delay its import for speeding up the common case.
+    if type(items) is list:
+        return items[:]
+    import copy
+    items = copy.copy(items)
```
Comment (wm75) [inline Lib/argparse.py]: shouldn't that be 'return copy.copy(items)' ?

Comment (serhiy-storchaka) [reply Lib/argparse.py]: Oh, right!

**Review by rhettinger — APPROVED** [review]

File: Lib/argparse.py [hunk 3]
```diff
@@ -137,10 +134,16 @@ def _get_args(self):
...
-    if getattr(namespace, name, None) is None:
-        setattr(namespace, name, value)
-    return getattr(namespace, name)
+def _copy_items(items):
+    if items is None:
+        return []
+    # The copy module is used only in the 'append' and 'append_const'
+    # actions, and it is needed only when the default value isn't a list.
+    # Delay its import for speeding up the common case.
+    if type(items) is list:
```
Comment (vstinner) [inline Lib/argparse.py:143]: Why not isinstance(items, list)?

File: Lib/argparse.py [hunk 4]
```diff
@@ -137,10 +134,16 @@ def _get_args(self):
...
+        return items[:]
```
Comment (vstinner) [inline Lib/argparse.py:144]: Why not items.copy()?

File: Lib/enum.py [hunk 5]
```diff
@@ -744,11 +742,10 @@ def __xor__(self, other):
...
         members, uncovered = _decompose(self.__class__, self._value_)
-        inverted_members = [
-                m for m in self.__class__
-                if m not in members and not m._value_ & self._value_
-                ]
-        inverted = reduce(_or_, inverted_members, self.__class__(0))
+        inverted = self.__class__(0)
+        for m in self.__class__:
+            if m not in members and not (m._value_ & self._value_):
+                inverted = inverted | m
```
Comment (vstinner) [inline Lib/enum.py:748]: I don't see how this change is related to removing imports.

File: Lib/gettext.py [hunk 6]
```diff
@@ -46,11 +46,9 @@
 #   find this format documented anywhere.
 
 
-import copy
 import locale
 import os
 import re
-import struct
 import sys
 from errno import ENOENT
```
Comment (vstinner) [inline Lib/gettext.py]: Would it be worth it to defer errno import as well? It's only used in one case, to raise an error when something goes wrong.

**Review by vstinner — COMMENTED** [review]
A few questions.

File: Lib/argparse.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/argparse.py:143]: `list` subclass can override `__copy__` or `__reduce__`.

File: Lib/argparse.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/argparse.py:144]: It is shorter.

File: Lib/enum.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/enum.py:748]: `functools.reduce` no longer imported.

File: Lib/gettext.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/gettext.py]: It is imported by `os` (where it is only used in one case too).

File: Lib/argparse.py (hunk 4 above)
Comment (rhettinger) [reply Lib/argparse.py:144]: And it's faster.

**Comment by serhiy-storchaka** [comment]
Got rid of importing `errno`.

**Review by methane — APPROVED** [review]
LGTM

File: Lib/argparse.py [hunk 7]
```diff
@@ -1000,8 +1010,10 @@ def __init__(self,
             help=help)
 
     def __call__(self, parser, namespace, values, option_string=None):
-        new_count = _ensure_value(namespace, self.dest, 0) + 1
-        setattr(namespace, self.dest, new_count)
+        count = getattr(namespace, self.dest, None)
+        if count is None:
+            count = 0
```
Comment (JimJJewett) [inline Lib/argparse.py:1015]: Why not just use getattr(namespace, self.dest, 0) and not have the if test?

Comment (merwok) [reply Lib/argparse.py:1015]: If think the current code is equivalent to `getattr(namespace, self.dest) or 0`, not `getattr(namespace, self.dest, 0)`.  The first one can never have None for count, the second does (if namespace has an attribute with the name in self.dest and the value None).

**Pull request merged**